### ISSUE TITLE
Launchpad's GHC 8.8.2 behaves differently from the official bindist

### ISSUE THREAD

**Issue by RyanGlScott** [haskell-CI/haskell-ci#353]
I originally reported this [here](https://gitlab.haskell.org/ghc/ghc/issues/17429#note_247282), but thought it important enough to bring to @hvr's attention here. It appears that the GHC 8.8.2 bindist that https://launchpad.net/~hvr/+archive/ubuntu/ghc offers behaves differently than the [official 8.8.2 bindist](https://www.haskell.org/ghc/download_ghc_8_8_2.html#linux_x86_64). You can observe the difference when compiling this program:

```hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE TypeFamilies #-}

module Core
  ( zoomAcceptor
  ) where

type Zoom m = ( m ~ Emitter Int )

zoomAcceptor :: Zoom m => Emitter w a -> m w
zoomAcceptor = fmap id . zoomEmitter

zoomEmitter :: Emitter w a -> Emitter b w
zoomEmitter (Emitter go) =
  Emitter $ const ([], fst $ go ())

newtype Emitter w a = Emitter (() -> ([w], [a]))

instance Functor (Emitter w) where
  fmap f (Emitter go) = Emitter mapped
    where
    {-# INLINE mapped #-}
    mapped _ = fmap f <$> go ()
```

When compiled with `-O -dcore-lint` using the Launchpad 8.8.2 bindist, it will work without issue. When compiled with `-O -dcore-lint` using the official 8.8.2 bindist, however, it will panic.

```
# Bindist from https://launchpad.net/~hvr/+archive/ubuntu/ghc
$ /opt/ghc/8.8.2/bin/ghc -O -dcore-lint -fforce-recomp Bug.hs
[1 of 1] Compiling Core             ( Bug.hs, Bug.o )

# Official bindist from https://www.haskell.org/ghc/download_ghc_8_8_2.html#linux_x86_64
$ ~/Software/sandbox/ghc-8.8.2/bin/ghc -O -dcore-lint -fforce-recomp Bug.hs
[1 of 1] Compiling Core             ( Bug.hs, Bug.o )
*** Core Lint errors : in result of CorePrep ***
<no location info>: warning:
    [in body of letrec with binders mapped_s1zR :: ()
                                                   -> ([Int], [w_s1w3])]
    Join point invoked with wrong number of arguments
    Join var: mapped_s1zR
    Join arity: 1
    Number of arguments: 0
```

This makes me wonder: are the two bindists built against the same commit? The official 8.8.2 bindist is tagged [here](https://gitlab.haskell.org/ghc/ghc/-/tags/ghc-8.8.2-release), which uses [this commit](https://gitlab.haskell.org/ghc/ghc/commit/86f4a56c8888d249194c38799eaec9484222fa47). On the other hand, the [`ghc-8.8` branch](https://gitlab.haskell.org/ghc/ghc/tree/ghc-8.8) currently contains two commits after this, [one of which is a fix](https://gitlab.haskell.org/ghc/ghc/commit/4d59e883b42a7f1fe56f1bf82aea826ebb9b1718) for the bug in the `Core` program above (and would explain the discrepancy).

**Comment by phadej** [comment]
```
% ghc-8.8.2 --print-project-git-commit-id
8b193877fe2843187f5984869549eb9643daebd3
```

I.e. not 
```
86f4a56c8888d249194c38799eaec9484222fa47
```

but indeed the one latter you posted.

**Comment by phadej** [comment]
BTW, it's probably the unfortunate consequence that
```
commit 8b193877fe2843187f5984869549eb9643daebd3
Author:     Tamar Christina <tamar@zhox.com>
AuthorDate: Fri Dec 20 23:05:57 2019 +0000
Commit:     Ben Gamari <ben@smart-cactus.org>
CommitDate: Wed Jan 15 14:41:46 2020 -0500

commit 4d59e883b42a7f1fe56f1bf82aea826ebb9b1718
Author:     Matheus Magalhães de Alcantara <matheus.de.alcantara@gmail.com>
AuthorDate: Tue Nov 19 16:39:47 2019 -0300
Commit:     Ben Gamari <ben@smart-cactus.org>
CommitDate: Wed Jan 15 13:49:53 2020 -0500
```

landed the 8.8.2 branch before the release announcement, as ghc-8.8.2 package was there tracking 8.8.2 branch already when it become apparent there will be one.

I guess @hvr should backtrack 8.8.2 and make 8.8.3 into the PPA tracking the ghc-8.8 branch tip.

**Comment by hvr** [comment]
I'm on it... turns `configure.ac` wasn't properly resetted to reflect a non-release state after the tag, so my scripting didn't catch this mistake and was happily mislead into believing this to be a proper release commit; i.e. the following two lines **MUST** be kept accurate for each commit on stable branches, or any automatic build-system will generate bindists which don't accurately reflect the version and screw-ups like this one happens... :-/

```
AC_INIT([The Glorious Glasgow Haskell Compilation System], [8.8.2], [glasgow-haskell-bugs@haskell.org], [ghc-AC_PACKAGE_VERSION])

# Set this to YES for a released version, otherwise NO
: ${RELEASE=YES}
```

**Comment by phadej** [comment]
I opened a GHC issue https://gitlab.haskell.org/ghc/ghc/issues/17709

**Comment by hvr** [comment]
The fixed up revisions of the `ghc-8.8.2` have almost finished building; I've already verified one of the binary packages that were already ready and they'll report the expected commit-id:

```
$ /opt/ghc/8.8.2/bin/ghc --info | grep Project
 [("Project name","The Glorious Glasgow Haskell Compilation System")
 ,("Project version","8.8.2")
 ,("Project Git commit id","86f4a56c8888d249194c38799eaec9484222fa47")

$ dpkg -l ghc-8.8.2
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                                                 Version                         Architecture                    Description
+++-====================================================-===============================-===============================-=============================================================================================================
ii  ghc-8.8.2                                            8.8.2-17~18.04                  amd64                           The Glasgow Haskell Compilation system

```

...and it also "features" the "wanted" bug triggering the `-dcore-lint` complaint.... ;-)

**Comment by RyanGlScott** [comment]
Thanks for the fix, @hvr! I can confirm that Launchpad's 8.8.2 bindists now point to the expected commit:

```
$ ghc --print-project-git-commit-id
86f4a56c8888d249194c38799eaec9484222fa47
```

I think the only thing left to do is on GHC's side, but that has already been reported https://gitlab.haskell.org/ghc/ghc/issues/17709. I'll go ahead and close this.
